### PR TITLE
Port -extra into API

### LINKED ISSUE #641: Migrate API-related utilities from `plutarch-extra` into ledger version specific library(#640)
Can be done after, or concurrently with #640. 

Move extra utilities into `plutarch-ledger-api-v...` as currently some utilities from `plutrach-extra` is ledger-verison-specific, and currently only support V1 ledger API

### PR DESCRIPTION
**Pull request by kozross**
Closes #641 .

### CODE REVIEW

**Review by SeungheonOh — DISMISSED** [review]
Everything looks good to me as most of it is copy from old code. We can merge once we dealt with `PTuple`

File: plutarch-api/src/Plutarch/Api.hs [hunk 1]
```diff
@@ -1195,30 +1199,124 @@ dataHash = hashData . Plutus.toData
...
+    PNothing -> pcon PNothing
+    PJust datumTuple -> pcon . PJust . ptryFromData . pforgetData $ psndBuiltin # pfromData datumTuple
+  where
+    matches ::
+      forall (s' :: S).
+      Term s' (PDatumHash :--> PAsData (PBuiltinPair (PAsData PDatumHash) (PAsData PDatum)) :--> PBool)
+    matches = phoistAcyclic $ plam $ \a ab -> a #== pfromData (pfstBuiltin # pfromData ab)
+
+{-
+pparseDatum = phoistAcyclic $
```
Comment (SeungheonOh) [inline plutarch-api/src/Plutarch/Api.hs:1304]: Clean this

File: plutarch-api/src/Plutarch/Api.hs [hunk 2]
```diff
@@ -1273,3 +1371,10 @@ pmaybeLT whenBothNothing ltF = phoistAcyclic $
             $ pconstant whenBothNothing
         )
       $ pconstant True
+
+ptryFromData ::
+  forall (a :: S -> Type) (s :: S).
+  PTryFrom PData (PAsData a) =>
+  Term s PData ->
+  Term s (PAsData a)
+ptryFromData x = unTermCont $ fst <$> tcont (ptryFrom @(PAsData a) x)
```
Comment (SeungheonOh) [inline plutarch-api/src/Plutarch/Api.hs:1380]: I don't think we need TermCont here

Comment (kozross) [reply plutarch-api/src/Plutarch/Api.hs:1380]: It's a direct copy. I'll rewrite it.

Comment (SeungheonOh) [reply plutarch-api/src/Plutarch/Api.hs:1380]: I don't think we need this helper really

**Comment by SeungheonOh** [comment]
Also, changed need to be logged now. I thought it was meant to overtake 640, but now that it's independent PR

**Comment by kozross** [comment]
> Also, changes need to be logged now. I thought it was meant to overtake 640, but now that it's independent PR

Happy to add those, but that means a version bump.

**Comment by SeungheonOh** [comment]
Yes, we need version bumped since we just deprecated a type

**Review by SeungheonOh — APPROVED** [review]
LGTM

**Pull request merged**